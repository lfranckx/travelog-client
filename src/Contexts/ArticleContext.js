/*eslint semi: ["error", "always"]*/
import React, { Component } from 'react';

export const nullArticle = {
    author: {},
    tags: [],
};
export const nullAuthor = {};
export const nullUser = {};
export const nullUsersArticles = [];

const ArticleContext = React.createContext({
    error: null,
    article: nullArticle,
    author: nullAuthor,
    user: nullUser,
    articlesList: [],
    authorsList: [],
    usersArticles: nullUsersArticles,
    setError: () => {},
    clearError: () => {},
    setArticle: () => {},
    clearArticle: () => {},
    setAuthor: () => {},
    clearAuthor: () => {},
    setUser: () => {},
    clearUser: () => {},
    setArticlesList: () => {},
    setAuthorsList: () => {},
    setUsersArticles: () => {},
    clearUsersArticles: () => {}
});
export default ArticleContext;

export class ArticleProvider extends Component {
    state = {
        error: null,
        article: nullArticle,
        author: nullAuthor,
        user: nullUser,
        articlesList: [],
        authorsList: [],
        usersArticles: []
    };

    setError = error => {
        console.error(error);
        this.setState({ error });
    };

    clearError = () => {
        this.setState({ error: null });
    };

    setArticle = article => {
        this.setState({ article });
    };

    clearArticle = () => {
        this.setArticle(nullArticle);
    };

    setAuthor = author => {
        this.setState({ author });
    };

    clearAuthor = () => {
        this.setAuthor(nullAuthor);
    };

    setUser = user => {
        console.log('setting user:', user);
        this.setState({ user });
    };

    clearUser = () => {
        this.setState(nullUser);
    };

    setArticlesList = articlesList => {
        this.setState({ articlesList });
    };

    setAuthorsList = authorsList => {
        this.setState({ authorsList });
    };

    setUsersArticles = usersArticles => {
        this.setState({ usersArticles });
    };

    clearUsersArticles = () => {
        this.setUsersArticles(nullUsersArticles);
    };

    render() {
        const value = {
            error: this.state.error,
            article: this.state.article,
            author: this.state.author,
            user: this.state.user,
            articlesList: this.state.articlesList,
            authorsList: this.state.authorsList,
            usersArticles: this.state.usersArticles,
            setError: this.setError,
            clearError: this.clearError,
            setArticle: this.setArticle,
            clearArticle: this.clearArticle,
            setAuthor: this.setAuthor,
            clearAuthor: this.clearAuthor,
            setUser: this.setUser,
            clearUser: this.clearUser,
            setArticlesList: this.setArticlesList,
            setAuthorsList: this.setAuthorsList,
            setUsersArticles: this.setUsersArticles,
            clearUsersArticles: this.clearUsersArticles
        };
        
        return (
            <ArticleContext.Provider value={value}>
                {this.props.children}
            </ArticleContext.Provider>
        );
    }   
}