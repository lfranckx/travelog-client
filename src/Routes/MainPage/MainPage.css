main {
    padding: 20px 180px;
}

main > section {
    text-align: left;
    padding: 20px 40px;
    border-radius: 2px;
}