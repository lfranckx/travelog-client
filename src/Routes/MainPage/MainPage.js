import React, { Component } from 'react';
import './MainPage.css';
import ArticlesListItem from '../../Components/ArticlesListItem/ArticlesListItem';
import ArticleContext from '../../Contexts/ArticleContext';
import ArticleApiService from '../../Services/article-api-service';
import AuthorApiService from '../../Services/author-api-service';
import TokenService from '../../Services/token-service';

class MainPage extends Component {
    static contextType = ArticleContext;

    componentDidMount() {
        this.context.clearError();
        ArticleApiService.getArticles()
            .then(this.context.setArticlesList)
            .catch(this.context.setError);
        AuthorApiService.getAuthors()
            .then(this.context.setAuthorsList)
            .catch(this.context.setError);
        if(TokenService.hasAuthToken()){
            AuthorApiService.getLoggedInAuthor()
                .then(this.context.setUser)
                .catch(this.context.setError);
        }
    }

    searchFor(term) {
        return function(x) {
            return x.first.toLowerCase().includes(term.toLowerCase()) || !term;
        };
    }

    renderArticles() {
        const { articlesList, authorsList } = this.context; 
        if (articlesList.length === 0) {
            return (
                <div className="loader">
                    <div className="spinner"></div>
                </div>
            ); 
        }
        
        // sort by date closest to current date
        articlesList.sort(function(a, b) {
            return new Date(b.date) - new Date(a.date);
        });

        // search filter by title
        let filteredList = articlesList.filter(article => {
            return article.title.toLowerCase().indexOf(
                this.context.search.toLowerCase()
            ) !== -1;
        });

        return filteredList.map(article => 
            <ArticlesListItem 
                key={article.id}
                article={article}
                authors={authorsList}
            />
        );
    }

    render() {
        const { error } = this.context;
        return (
            <>
                <section className="main-page-articles">
                    <div>
                        <h2>Share authentic travel stories</h2>
                        <h3 id="sub">Connect with other travelers and inspire</h3>
                    </div>
                    <h3 id="stories">Stories,</h3>
                    {error
                        ? <p className='error' >There was an error try again</p>
                        : this.renderArticles()
                    }
                </section>
            </>
        );
    }
}

export default MainPage;