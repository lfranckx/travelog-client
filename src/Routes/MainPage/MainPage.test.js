import React from 'react';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import MainPage from './MainPage';
import { ArticleProvider } from '../../Contexts/ArticleContext';

describe(`MainPage component`, () => {
    it(`renders page without crashing`, () => {
        const wrapper = mount(
            <ArticleProvider>
                <MainPage />
            </ArticleProvider>
        )
        expect(toJson(wrapper)).toMatchSnapshot();
    });
});