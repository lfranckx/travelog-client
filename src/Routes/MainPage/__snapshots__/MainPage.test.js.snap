// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainPage component renders page without crashing 1`] = `
<ArticleProvider>
  <MainPage>
    <section
      className="main-page-articles"
    >
      <h2>
        Stories,
      </h2>
      <div
        className="loader"
      >
        <div
          className="spinner"
        />
      </div>
    </section>
  </MainPage>
</ArticleProvider>
`;
